<a href="https://colab.research.google.com/github/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/blob/main/Python%20NoteBooks/Regresi%C3%B3n_lineal_m%C3%BAltiple_Lasso_y_Ridge_calidad_de_servicio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Contexto de los datos

Al igual que el caso de R del punto anterior, se trata de un conjunto de datos en la que se valora la calidad de servicio de una empresa u organziación que recopila datos en relación a esta calidad.

Los datos que se utilzian se encuentran en el enlace de *github.com* https://raw.githubusercontent.com/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/refs/heads/main/datos/calidad%20en%20el%20servicio%20para%20Multiple%20Lasso%20y%20Ridge.csv .

Luego el archivo se encuentra para su reproduccón es en enlace de *goole collab* https://colab.research.google.com/drive/1-VnGQfBhydoDnQu8tnLC2DMefA5vakL8?usp=sharing .

De igual forma el archivo se encuntra para su descarga direcamente en el serbicio github.com en el espacio del autor: https://github.com/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/blob/main/Python%20NoteBooks/Regresi%C3%B3n_lineal_m%C3%BAltiple_Lasso_y_Ridge_calidad_de_servicio.ipynb .


Las funciones previamente programadas y preparadas se encuentran en *github.com*; https://raw.githubusercontent.com/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/refs/heads/main/Python%20NoteBooks/funciones/Funciones%20para%20implementar%20y%20evaluar%20modelos%20regresion%20lineal%20multiple%20ademas%20de%20Lasso%20y%20Ridge.py .

Son alrededor de *2000* registros. Todas las variables del conjunto de datos son de tipo numérico, por lo que no es necesario algún tipo de preparación de variables o transformarlas; estas son las variables independientes.

* *tiempo_respuesta*: Tiempo promedio (en unidades de tiempo) que tarda la organización en atender una solicitud o requerimiento del cliente. A menor tiempo, mayor calidad esperada.
* *satisfaccion_cliente*: Nivel promedio de satisfacción del cliente respecto al servicio recibido; variable atributo directamente relacionado con la calidad.
* *numero_quejas*: Número promedio de quejas registradas por parte de los clientes; a mayor número de quejas, menor calidad del servicio.
* *capacitacion_empleado*: Nivel promedio de capacitación del personal que brinda el servicio. A mayor capacitación, mejor desempeño.
* *experiencia_empleado*: Años o nivel de experiencia promedio del personal; contribuye a la eficiencia y calidad del servicio.
* disponibilidad_servicio: Grado en que el servicio está disponible cuando el cliente lo requiere; alta disponibilidad mejora la percepción de calidad.
* tiempo_respuesta_2: Variable derivada altamente correlacionada con tiempo_respuesta y significa mediciones alternativas o redundantes del mismo fenómeno.
* *satisfaccion_2*: Variable altamente correlacionada con satisfaccion_cliente; simula escalas distintas de medición del mismo concepto.
* indice_servicio: Indicador compuesto que combina satisfacción y disponibilidad del servicio; representa una medida agregada del desempeño.
* *eficiencia_operativa*: Indicador que combina rapidez en respuesta y experiencia del personal; refleja la eficiencia interna del servicio.
* *clima*: Condiciones climáticas promedio durante el periodo de evaluación; tal vez pueda influir indirectamente en el servicio.
* *dia_semana*: Día de la semana en que se brinda el servicio; representa variaciones operativas demanda, carga de trabajo, entre otros.

Se percibe que en este conjunto de datos, existen variable relevantes como *tiempo_respuesta*, *satisfaccion_cliente*, *numero_quejas*, *capacitacion_empleado*, disponibilidad_servico; tal vez existan variables muy correlacionadas y/o colineales como *tiempo_respuesta_2*, *satisfaccion_2*; es posible que existan variables derivadas de otras por lo que puedan estar correlacioandas *indice_servicio*, *eficiencia_operativa* y es probable que existan variables poco deterministas para las predicción de la calidad de servicio como *clima*, *dia_semana*.


La variable dependiente es *calidad_servicio* como un indicador global de un valor numérico posiblemente encuestado y calculado que identifica la calidad del servicio percibida por el cliente; integra múltiples factores como eficiencia, atención, disponibilidad y experiencia del servicio, entre otros.



# Objetivo

Implementar, evaluar un modelo de regresión lineal múltiple así como compararlo con los modelos generalizados y regulados como *Lasso* y *Ridge* que permitan predecir calidad en el servico de una organización.

Los datos deberán estar particionados en *70%* para datos de entrenamiento y *30%* para datos de validación

El modelo será óptimo si satisface los postulados de la regresión y si el valor de *r square adjustado* está por encima del *85%*.


# Descripción

## Cargar librerías



In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LassoCV
from sklearn.linear_model import RidgeCV
from sklearn.metrics import (
    mean_squared_error,
    mean_absolute_error,
    r2_score
)

from scipy.stats import shapiro
from scipy.stats import kstest
from statsmodels.stats.diagnostic import het_breuschpagan
from statsmodels.stats.stattools import durbin_watson
from statsmodels.stats.outliers_influence import variance_inflation_factor
from statsmodels.stats.diagnostic import linear_reset
import statsmodels.api as sm

# Para validar posutalados
from statsmodels.stats.outliers_influence import (
    variance_inflation_factor
)
from statsmodels.stats.diagnostic import (
    het_breuschpagan,
    linear_reset
)
from statsmodels.stats.stattools import (
    durbin_watson
)
from scipy.stats import (
    shapiro
)

from scipy.stats import shapiro
from scipy.stats import kstest
from scipy.stats import anderson

from sklearn.linear_model import (
    LassoCV
)

from sklearn.linear_model import (
    RidgeCV
)


## Cargar funciones


Se cargan las funciones que se encuentra en el portal de *github.com*; https://raw.githubusercontent.com/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/refs/heads/main/Python%20NoteBooks/funciones/Funciones%20para%20implementar%20y%20evaluar%20modelos%20regresion%20lineal%20multiple%20ademas%20de%20Lasso%20y%20Ridge.py


In [26]:
# Aqui se carga
import requests

url = "https://raw.githubusercontent.com/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/refs/heads/main/Python%20NoteBooks/funciones/Funciones%20para%20implementar%20y%20evaluar%20modelos%20regresion%20lineal%20multiple%20ademas%20de%20Lasso%20y%20Ridge.py"

exec(requests.get(url).text)

## Cargar datos



In [27]:
url = "https://raw.githubusercontent.com/rpizarrog/Libro-Aprendizaje-Automatico.-Casos-de-Estudio-con-R-y-Python/refs/heads/main/datos/calidad%20en%20el%20servicio%20para%20Multiple%20Lasso%20y%20Ridge.csv"
datos = f_cargar_datos(url)


## descripci[on de datos


In [28]:
f_describir_datos(datos)

{'describe':        tiempo_respuesta  satisfaccion_cliente  numero_quejas  \
 count       2000.000000           2000.000000    2000.000000   
 mean           9.997575              6.983065       5.018290   
 std            1.957297              1.001619       2.040416   
 min            3.900000              3.330000      -1.680000   
 25%            8.670000              6.290000       3.700000   
 50%           10.005000              6.970000       5.040000   
 75%           11.240000              7.660000       6.380000   
 max           16.830000             10.090000      12.480000   
 
        capacitacion_empleado  experiencia_empleado  disponibilidad_servicio  \
 count            2000.000000           2000.000000              2000.000000   
 mean                6.040910              8.000355                 9.010175   
 std                 1.507774              2.003330                 1.008342   
 min                 1.320000              1.360000                 5.780000   
 

## Visualizar datos


In [29]:
f_visualizar_head_tail_reducido(datos)

,tiempo_respuesta,satisfaccion_cliente,numero_quejas,capacitacion_empleado,...,eficiencia_operativa,clima,dia_semana,calidad_servicio
0,11.04,7.2,7.47,6.16,...,0.58,12.46,3,102.57
1,7.84,7.16,4.35,5.19,...,2.17,25.72,4,112.76
2,10.28,7.08,8.71,8.41,...,2.41,21.02,1,105.95
3,9.83,6.17,1.78,8.1,...,0.12,27.17,5,102.94
4,8.67,6.38,3.45,5.63,...,1.18,22.09,6,103.39
5,4.97,7.6,6.04,6.26,...,2.99,28.86,6,114.32
6,...,...,...,...,...,...,...,...,...
7,8.39,5.77,5.22,4.95,...,0.58,23.37,4,101.06
8,8.09,7.87,4.78,5.51,...,1.02,32.81,5,104.06
9,6.82,6.1,7.83,6.8,...,2.28,30.62,5,102.35


# Desarrollo

Se particionan los datos, se construye el modelo de regresión lineal múltiple, se validan los postulados y se identifican deficiencias en el modelo.

## Particionar los datos

Se generan los dos conjuntos de datos, *70%* para datos de entrenamiento y *30%* para datos de validación.


In [30]:
particion = f_particionar_datos(datos)

datos_entrenamiento = particion["datos_entrenamiento"]
datos_validacion = particion["datos_validacion"]

f_visualizar_head_tail_reducido(datos_entrenamiento)


,tiempo_respuesta,satisfaccion_cliente,numero_quejas,capacitacion_empleado,...,eficiencia_operativa,clima,dia_semana,calidad_servicio
0,13.05,7.98,3.96,4.48,...,0.52,19.95,4,103.44
1,7.72,7.7,4.72,5.94,...,2.67,26.83,3,108.63
2,7.8,6.99,1.99,5.4,...,1.74,26.94,6,103.15
3,10.87,6.44,5.23,8.37,...,-0.33,22.66,5,99.19
4,9.5,4.64,4.79,8.79,...,3.63,24.81,6,109.45
5,8.85,8.25,5.99,5.43,...,4.13,19.98,2,117.9
6,...,...,...,...,...,...,...,...,...
7,8.2,5.95,4.96,7.38,...,-0.11,25.6,5,99.97
8,11.38,5.51,2.43,5.17,...,-1.05,16.71,3,90.78
9,10.03,7.23,6.72,5.4,...,0.49,25.96,5,91.91


In [31]:
f_visualizar_head_tail_reducido(datos_validacion)

,tiempo_respuesta,satisfaccion_cliente,numero_quejas,capacitacion_empleado,...,eficiencia_operativa,clima,dia_semana,calidad_servicio
0,13.08,5.85,8.46,6.05,...,0.9,25.34,4,90.49
1,6.81,7.2,8.49,4.9,...,0.32,30.12,6,90.2
2,9.15,8.38,7.68,6.87,...,0.38,22.13,2,99.21
3,9.09,7.36,3.62,5.93,...,1.14,10.56,2,99.86
4,7.66,6.4,6.24,3.7,...,1.68,25.29,6,95.07
5,8.43,6.82,6.35,6.33,...,2.18,31.43,5,107.73
6,...,...,...,...,...,...,...,...,...
7,11.39,5.9,5.68,10.5,...,0.29,24.4,5,106.52
8,7.71,8.25,4.97,8.09,...,1.85,17.93,7,106.94
9,10.45,7.39,4.31,3.54,...,1.2,23.09,7,103.38


## Construir modelos de regresión lineal múltiple

Se construye el modelo de regresión lineal múltiple mandando llamar la función *f_construir_modelo_RLM()*:

In [8]:
modelo_RLM = f_construir_modelo_RLM_statsmodels(datos_entrenamiento, "calidad_servicio")

                            OLS Regression Results                            
Dep. Variable:       calidad_servicio   R-squared:                       0.894
Model:                            OLS   Adj. R-squared:                  0.893
Method:                 Least Squares   F-statistic:                     976.8
Date:                Mon, 18 May 2026   Prob (F-statistic):               0.00
Time:                        16:56:29   Log-Likelihood:                -3529.8
No. Observations:                1400   AIC:                             7086.
Df Residuals:                    1387   BIC:                             7154.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                              coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------
const                     

## Postulados del modelo

Mediante la función *f_validar_postulados_modelo()* se devuelve en modo consola la condición de cada postulado; el diagnóstico indica que el valor de *VIF* aproximadamente de *18.88* indica que el modelo no cumpla con el postulado de multicolinealidad y que seguramente hay variables muy correlacionadas entre si; por otra parte si se cumplen los demás postulados como lo indica el resultado de esta función que son similares al caso de estudio de descrito en *R*.

Como no se cumple con el postulado de multicolinealidad, se realizan dos estrategias: 1. combinar variables, 2. Construir modelos Lasso y Ridge





In [10]:
f_validar_postulados_modelos (modelo_RLM)


Validación de Postulados
     Modelo Multicolinealidad  VIF_Max Linealidad Homocedasticidad Normalidad  \
0  Modelo_1         No cumple  18.8895     Cumple           Cumple     Cumple   

  Independencia  
0        Cumple  


,Modelo,Multicolinealidad,VIF_Max,Linealidad,Homocedasticidad,Normalidad,Independencia
0,Modelo_1,No cumple,18.8895,Cumple,Cumple,Cumple,Cumple


## Evaluar modelo


In [9]:
f_evaluacion_modelos(modelo_RLM, datos_validacion, "calidad_servicio")


Evaluación de Modelos
     Modelo  R_square  R_square_ajustado     MSE    RMSE     MAE
0  Modelo_1    0.8952             0.8931  8.9634  2.9939  2.4066


,Modelo,R_square,R_square_ajustado,MSE,RMSE,MAE
0,Modelo_1,0.8952,0.8931,8.9634,2.9939,2.4066


## Estrategia 1 para disminuir multicolinealidad. Modelo de regresión múltiple con datos depurados

Se construye otro modelo lineal múltiple con variables combinadas y depuradas; se sugiere combinar (sumar los valores) *tiempo_respuesta* y *tiempo_respuesta_2* en una sola variable llamada *tiempo_respuesta_indicador*, luego combinar (sumar los valores) *satisfaccion_cliente* con *satisfaccion_cliente_2* y dejarlo en una variable única llamada *satisfaccion_cliente_indicador*, y dejar la variable *eficiencia_operativa* como está.

El nuevo conjunto de datos se llama *datos_depurados*.

In [11]:
# DEPURACIÓN / SELECCIÓN DE VARIABLES
datos_depurados = pd.DataFrame({
    "tiempo_respuesta_indicador":
        datos["tiempo_respuesta"] + datos["tiempo_respuesta_2"],
    "satisfaccion_cliente_indicador":
        datos["satisfaccion_cliente"] + datos["satisfaccion_2"],
    "numero_quejas":
        datos["numero_quejas"],
    "capacitacion_empleado":
        datos["capacitacion_empleado"],
    "experiencia_empleado":
        datos["experiencia_empleado"],

    "disponibilidad_servicio":
        datos["disponibilidad_servicio"],
    "indice_servicio":
        datos["indice_servicio"],
    "eficiencia_operativa":
        datos["eficiencia_operativa"],
    "clima":
        datos["clima"],
    "dia_semana":
        datos["dia_semana"],
    "calidad_servicio":
        datos["calidad_servicio"]
})

datos_depurados = datos_depurados.round(4)

f_visualizar_head_tail_reducido(datos_depurados)

,tiempo_respuesta_indicador,satisfaccion_cliente_indicador,numero_quejas,capacitacion_empleado,...,eficiencia_operativa,clima,dia_semana,calidad_servicio
0,21.97,13.7,7.47,6.16,...,0.58,12.46,3,102.57
1,15.38,13.45,4.35,5.19,...,2.17,25.72,4,112.76
2,20.38,13.38,8.71,8.41,...,2.41,21.02,1,105.95
3,19.87,11.67,1.78,8.1,...,0.12,27.17,5,102.94
4,16.89,12.42,3.45,5.63,...,1.18,22.09,6,103.39
5,10.1,14.85,6.04,6.26,...,2.99,28.86,6,114.32
6,...,...,...,...,...,...,...,...,...
7,16.15,10.76,5.22,4.95,...,0.58,23.37,4,101.06
8,16.08,14.81,4.78,5.51,...,1.02,32.81,5,104.06
9,12.63,11.61,7.83,6.8,...,2.28,30.62,5,102.35


### Datos de entenamiento y validación de los datos depurados


In [32]:
particion = f_particionar_datos(datos_depurados)

datos_entrenamiento = particion["datos_entrenamiento"]
datos_validacion = particion["datos_validacion"]

f_visualizar_head_tail_reducido(datos_entrenamiento)

,tiempo_respuesta_indicador,satisfaccion_cliente_indicador,numero_quejas,capacitacion_empleado,...,eficiencia_operativa,clima,dia_semana,calidad_servicio
0,25.62,14.95,3.96,4.48,...,0.52,19.95,4,103.44
1,15.46,15.36,4.72,5.94,...,2.67,26.83,3,108.63
2,15.95,13.2,1.99,5.4,...,1.74,26.94,6,103.15
3,22.3,12.03,5.23,8.37,...,-0.33,22.66,5,99.19
4,18.24,8.31,4.79,8.79,...,3.63,24.81,6,109.45
5,17.52,15.3,5.99,5.43,...,4.13,19.98,2,117.9
6,...,...,...,...,...,...,...,...,...
7,16.02,11.4,4.96,7.38,...,-0.11,25.6,5,99.97
8,22.23,10.77,2.43,5.17,...,-1.05,16.71,3,90.78
9,20.17,13.41,6.72,5.4,...,0.49,25.96,5,91.91


In [13]:
f_visualizar_head_tail_reducido(datos_validacion)

,tiempo_respuesta_indicador,satisfaccion_cliente_indicador,numero_quejas,capacitacion_empleado,...,eficiencia_operativa,clima,dia_semana,calidad_servicio
0,26.17,11.15,8.46,6.05,...,0.9,25.34,4,90.49
1,13.49,13.79,8.49,4.9,...,0.32,30.12,6,90.2
2,19.08,15.82,7.68,6.87,...,0.38,22.13,2,99.21
3,18.81,13.63,3.62,5.93,...,1.14,10.56,2,99.86
4,14.91,12.49,6.24,3.7,...,1.68,25.29,6,95.07
5,16.53,13.05,6.35,6.33,...,2.18,31.43,5,107.73
6,...,...,...,...,...,...,...,...,...
7,22.54,11.35,5.68,10.5,...,0.29,24.4,5,106.52
8,16.44,15.51,4.97,8.09,...,1.85,17.93,7,106.94
9,21.13,14.5,4.31,3.54,...,1.2,23.09,7,103.38


##

###  Construir modelo de regresión lineal múltiple con datos depurados

Se construye el modelo, pero con los datos de entrenamiento que han sido depurados.


In [14]:
modelo_RLM_depurado = f_construir_modelo_RLM_statsmodels(datos_entrenamiento, "calidad_servicio")

                            OLS Regression Results                            
Dep. Variable:       calidad_servicio   R-squared:                       0.891
Model:                            OLS   Adj. R-squared:                  0.890
Method:                 Least Squares   F-statistic:                     1133.
Date:                Mon, 18 May 2026   Prob (F-statistic):               0.00
Time:                        16:56:30   Log-Likelihood:                -3552.0
No. Observations:                1400   AIC:                             7126.
Df Residuals:                    1389   BIC:                             7184.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                                     coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
const       

### Validar postulados de modelos

Se validan los postulados tanto del modelo de regresión múltiple con datos originales como del modelo de regresión múltiple con datos depurados.

In [15]:
f_validar_postulados_modelos([modelo_RLM,modelo_RLM_depurado],
    nombres = ["RLM Original","RLM Depurado"]
)


Validación de Postulados
         Modelo Multicolinealidad  VIF_Max Linealidad Homocedasticidad  \
0  RLM Original         No cumple  18.8895     Cumple           Cumple   
1  RLM Depurado          Moderada   8.4910     Cumple           Cumple   

  Normalidad Independencia  
0     Cumple        Cumple  
1     Cumple        Cumple  


,Modelo,Multicolinealidad,VIF_Max,Linealidad,Homocedasticidad,Normalidad,Independencia
0,RLM Original,No cumple,18.8895,Cumple,Cumple,Cumple,Cumple
1,RLM Depurado,Moderada,8.4910,Cumple,Cumple,Cumple,Cumple


In [16]:

f_evaluacion_modelos([modelo_RLM_depurado],
    datos_validacion,
    "calidad_servicio",
    nombres = ["RLM Depurado"]
)


Evaluación de Modelos
         Modelo  R_square  R_square_ajustado     MSE    RMSE     MAE
0  RLM Depurado    0.8919             0.8901  9.2426  3.0402  2.4299


,Modelo,R_square,R_square_ajustado,MSE,RMSE,MAE
0,RLM Depurado,0.8919,0.8901,9.2426,3.0402,2.4299


### Estrategia 2. Construir modelos regresión lineal múltiple, Lasso y Ridge con datos estandarizados

Antes de construir los modelos regresión lineal múltiple, *Lasso* y *Ridge* respectivamente, se sugiere estandarizar o escalar los datos originales.

Se ejecuta la función *f_estandarizar_escalar()* la cual devuelve datos estandarizados y escalados a cuatro posiciones decimales.

Para la construcción de  los modelos *Lasso* y *Ridge* se emplean únicamente los datos estandarizados.

Se observan los primeros y últimos registros de los datos de entrenamiento y validación de los datos estandarizados.



In [17]:
datos_estandarizados_escalados = f_estandarizar_escalar(datos)
datos_estandarizados = datos_estandarizados_escalados['datos_estandarizados']
datos_escalados = datos_estandarizados_escalados['datos_escalados']

f_visualizar_head_tail_reducido(datos_estandarizados)

,tiempo_respuesta,satisfaccion_cliente,numero_quejas,capacitacion_empleado,...,eficiencia_operativa,clima,dia_semana,calidad_servicio
0,0.5327,0.2166,1.2019,0.079,...,-0.1518,-2.4893,-0.5107,-0.0061
1,-1.1026,0.1767,-0.3276,-0.5645,...,0.8962,0.1265,-0.0095,1.0949
2,0.1443,0.0968,1.8097,1.5716,...,1.0544,-0.8006,-1.5132,0.3591
3,-0.0856,-0.812,-1.5875,1.366,...,-0.455,0.4126,0.4917,0.0338
4,-0.6784,-0.6022,-0.7688,-0.2726,...,0.2437,-0.5895,0.9929,0.0825
5,-2.5693,0.6161,0.5009,0.1453,...,1.4367,0.746,0.9929,1.2634
6,...,...,...,...,...,...,...,...,...
7,-0.8215,-1.2114,0.0989,-0.7237,...,-0.1518,-0.337,-0.0095,-0.1693
8,-0.9748,0.8857,-0.1168,-0.3522,...,0.1382,1.5252,0.4917,0.1549
9,-1.6239,-0.8819,1.3784,0.5036,...,0.9687,1.0932,0.4917,-0.0299


### Particionar datos estandarizados

Ahora nuevas particiones pero con datos estandarizados


In [18]:
particion = f_particionar_datos(datos_estandarizados)

datos_entrenamiento = particion["datos_entrenamiento"]
datos_validacion = particion["datos_validacion"]

f_visualizar_head_tail_reducido(datos_entrenamiento)

,tiempo_respuesta,satisfaccion_cliente,numero_quejas,capacitacion_empleado,...,eficiencia_operativa,clima,dia_semana,calidad_servicio
0,1.5599,0.9956,-0.5188,-1.0355,...,-0.1914,-1.0117,-0.0095,0.0879
1,-1.1639,0.716,-0.1462,-0.0669,...,1.2258,0.3455,-0.5107,0.6486
2,-1.123,0.0069,-1.4845,-0.4252,...,0.6128,0.3672,0.9929,0.0565
3,0.4458,-0.5423,0.1038,1.5451,...,-0.7517,-0.4771,0.4917,-0.3713
4,-0.2543,-2.3399,-0.1119,1.8237,...,1.8586,-0.053,0.9929,0.7372
5,-0.5865,1.2652,0.4764,-0.4053,...,2.1882,-1.0058,-1.012,1.6502
6,...,...,...,...,...,...,...,...,...
7,-0.9186,-1.0317,-0.0286,0.8883,...,-0.6066,0.1029,0.4917,-0.287
8,0.7065,-1.4711,-1.2688,-0.5778,...,-1.2263,-1.6509,-0.5107,-1.28
9,0.0166,0.2466,0.8342,-0.4252,...,-0.2112,0.1739,0.4917,-1.1579


In [19]:
f_visualizar_head_tail_reducido(datos_validacion)

,tiempo_respuesta,satisfaccion_cliente,numero_quejas,capacitacion_empleado,...,eficiencia_operativa,clima,dia_semana,calidad_servicio
0,1.5752,-1.1315,1.6872,0.006,...,0.0591,0.0516,-0.0095,-1.3113
1,-1.629,0.2166,1.7019,-0.7569,...,-0.3232,0.9945,0.9929,-1.3427
2,-0.4331,1.395,1.3048,0.55,...,-0.2837,-0.5817,-1.012,-0.3692
3,-0.4638,0.3764,-0.6855,-0.0736,...,0.2173,-2.8641,-1.012,-0.2989
4,-1.1946,-0.5823,0.5989,-1.5529,...,0.5732,0.0417,0.9929,-0.8165
5,-0.8011,-0.1628,0.6528,0.1918,...,0.9028,1.253,0.4917,0.5514
6,...,...,...,...,...,...,...,...,...
7,0.7116,-1.0816,0.3244,2.9581,...,-0.343,-0.1338,0.4917,0.4206
8,-1.169,1.2652,-0.0237,1.3594,...,0.6853,-1.4102,1.4941,0.466
9,0.2312,0.4064,-0.3472,-1.6591,...,0.2568,-0.3923,1.4941,0.0814


### Construir modelo regresión lineal múltiple con datos erstandarizdos


In [20]:
modelo_RLM_estandarizado = f_construir_modelo_RLM_statsmodels(datos_entrenamiento, "calidad_servicio")

                            OLS Regression Results                            
Dep. Variable:       calidad_servicio   R-squared:                       0.894
Model:                            OLS   Adj. R-squared:                  0.893
Method:                 Least Squares   F-statistic:                     976.8
Date:                Mon, 18 May 2026   Prob (F-statistic):               0.00
Time:                        16:56:31   Log-Likelihood:                -414.55
No. Observations:                1400   AIC:                             855.1
Df Residuals:                    1387   BIC:                             923.3
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
                              coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------
const                     

### Construir modelo Lasso



In [21]:
modelo_Lasso = f_construir_modelo_lasso(datos_entrenamiento, "calidad_servicio")


MODELO LASSO
Lambda óptimo: 0.001
Intercepto: -0.0085

Coeficientes:
                   Variable  Coeficiente
0          tiempo_respuesta      -0.3465
1      satisfaccion_cliente       0.3071
2             numero_quejas      -0.3214
3     capacitacion_empleado       0.3074
4      experiencia_empleado       0.5198
5   disponibilidad_servicio       0.3211
6        tiempo_respuesta_2      -0.0522
7            satisfaccion_2       0.0177
8           indice_servicio      -0.0000
9      eficiencia_operativa       0.0212
10                    clima      -0.0062
11               dia_semana      -0.0081


### Construir modelo Ridge

In [22]:
modelo_Ridge = f_construir_modelo_ridge(datos_entrenamiento, "calidad_servicio")


MODELO RIDGE
Lambda óptimo: 1.9179
Intercepto: -0.0085

Coeficientes:
                   Variable  Coeficiente
0          tiempo_respuesta      -0.3380
1      satisfaccion_cliente       0.3059
2             numero_quejas      -0.3218
3     capacitacion_empleado       0.3079
4      experiencia_empleado       0.5167
5   disponibilidad_servicio       0.3235
6        tiempo_respuesta_2      -0.0594
7            satisfaccion_2       0.0214
8           indice_servicio      -0.0033
9      eficiencia_operativa       0.0255
10                    clima      -0.0072
11               dia_semana      -0.0091


### Validar postulados

Se validan todos los postulados, tomando en cuenta que los modelos *Lasso* y *Ridge* no validan multicolinealidad dado que ya son modelos regularizados.

Al ejecutar nuevamente la función *f_validar_modelos()*, el resultado en modo consola indica que aún existe multicolinealidad en los tres modelos que utilizan datos estandarizados; incluyendo *Lasso* y *Ridge*.

Lo anterior, se debe interpretar que las correlación sigue existiendo entre las variable independientes, solo que los modelos *Lasso* y *Ridge* no eliminan la multicolinealidad presente entre las variables independientes, ya que el *VIF* depende de la relación entre las variables predictoras y no del algoritmo de regresión utilizado.

Los modelos *Lasso* y *Ridge* como se indicó en el fundamento teórico arriba descrito, incorporan mecanismos de regularización que permiten disminuir los efectos negativos de la colinealidad sobre la estabilidad y magnitud de los coeficientes. En lo particular, *Ridge* ser más robusto ante problemas severos de multicolinealidad debido a la penalización cuadrática aplicada sobre los coeficientes del modelo.


In [23]:
X_est = datos_estandarizados.drop(columns = ["calidad_servicio"]) # Variables independientes

y_est = datos_estandarizados["calidad_servicio"] # Variable dependiente

f_validar_postulados_modelos(
    [modelo_RLM_estandarizado, modelo_Lasso, modelo_Ridge],
    X_datos = X_est,
    y_datos = y_est,
    nombres = ["RLM", "LASSO", "RIDGE"]
)


Validación de Postulados
  Modelo Multicolinealidad  VIF_Max Linealidad Homocedasticidad Normalidad  \
0    RLM         No cumple  18.8895     Cumple           Cumple     Cumple   
1  LASSO         No cumple  18.8530         NA               NA     Cumple   
2  RIDGE         No cumple  18.8530         NA               NA     Cumple   

  Independencia  
0        Cumple  
1        Cumple  
2        Cumple  


,Modelo,Multicolinealidad,VIF_Max,Linealidad,Homocedasticidad,Normalidad,Independencia
0,RLM,No cumple,18.8895,Cumple,Cumple,Cumple,Cumple
1,LASSO,No cumple,18.8530,NA,NA,Cumple,Cumple
2,RIDGE,No cumple,18.8530,NA,NA,Cumple,Cumple


## Evaluación del modelo


In [24]:
f_evaluacion_modelos([modelo_RLM_estandarizado, modelo_Lasso, modelo_Ridge],
    datos_validacion,
    "calidad_servicio",
    nombres = ["RLM", "LASSO", "RIDGE"]
)


Evaluación de Modelos
  Modelo  R_square  R_square_ajustado     MSE    RMSE     MAE
0    RLM    0.8952             0.8931  0.1046  0.3235  0.2600
1  LASSO    0.8952             0.8931  0.1046  0.3235  0.2600
2  RIDGE    0.8951             0.8930  0.1047  0.3236  0.2601


,Modelo,R_square,R_square_ajustado,MSE,RMSE,MAE
0,RLM,0.8952,0.8931,0.1046,0.3235,0.2600
1,LASSO,0.8952,0.8931,0.1046,0.3235,0.2600
2,RIDGE,0.8951,0.8930,0.1047,0.3236,0.2601


### Ecuaciones de modelos

Finalmente llamando a la función *f_ecuaciones_modelos()* se muestran en modo cnsola las ecuacioesn que definen las predicciones de cada modelo construido.



In [25]:
f_ecuaciones_modelos([modelo_RLM, modelo_RLM_depurado,
                      modelo_RLM_estandarizado, modelo_Lasso, modelo_Ridge],
    nombres = ["RLM Original", "RLM Depurado", "RLM Estandarizado", "LASSO", "RIDGE"])


ECUACIÓN: RLM Original

ŷ = 50.8175 - 1.6461*tiempo_respuesta + 2.8703*satisfaccion_cliente - 1.4623*numero_quejas + 1.8927*capacitacion_empleado + 2.4166*experiencia_empleado + 2.983*disponibilidad_servicio - 0.2434*tiempo_respuesta_2 + 0.1742*satisfaccion_2 - 0.0521*indice_servicio + 0.1137*eficiencia_operativa - 0.013*clima - 0.0429*dia_semana

ECUACIÓN: RLM Depurado

ŷ = 51.0384 - 0.911*tiempo_respuesta_indicador + 1.518*satisfaccion_cliente_indicador - 1.4518*numero_quejas + 1.8969*capacitacion_empleado + 2.3567*experiencia_empleado + 2.8988*disponibilidad_servicio + 0.0814*indice_servicio + 0.2165*eficiencia_operativa - 0.0118*clima - 0.0225*dia_semana

ECUACIÓN: RLM Estandarizado

ŷ = -0.0085 - 0.348*tiempo_respuesta + 0.3105*satisfaccion_cliente - 0.3223*numero_quejas + 0.3083*capacitacion_empleado + 0.5229*experiencia_empleado + 0.3249*disponibilidad_servicio - 0.0533*tiempo_respuesta_2 + 0.018*satisfaccion_2 - 0.005*indice_servicio + 0.0186*eficiencia_operativa - 0.0071*clim

,Modelo,Ecuacion
0,RLM Original,ŷ = 50.8175 - 1.6461*tiempo_respuesta + 2.8703...
1,RLM Depurado,ŷ = 51.0384 - 0.911*tiempo_respuesta_indicador...
2,RLM Estandarizado,ŷ = -0.0085 - 0.348*tiempo_respuesta + 0.3105*...
3,LASSO,ŷ = -0.0085 - 0.3465*tiempo_respuesta + 0.3071...
4,RIDGE,ŷ = -0.0085 - 0.338*tiempo_respuesta + 0.3059*...


# Interpretación del caso de estudio

En este caso de estudio, se construyeron varios modelos de regresión.

Primero se construyó un modelo de regresión lineal múltiple con los datos originales; al validarse los postulados, se detecta una multicolinealidad en las variables independientes por lo que el modelo adolece de esta condición, sin embargo, el modelo cumple con la expectativa de calidad predictiva, en la evaluación del modelo el estadístico r square ajustado es de 0.8931 por lo que cumple con expectativa de estar por encima 85% establecido como métrica de valoración.

Para disminuir multicolinealidad se siguió una primera estrategia, construyendo un nuevo modelo, pero con variables depuradas, aquellas variables que fueron detectadas muy correlacionadas se juntaron lo que permitió tener un nuevo conjunto de datos. Al evaluar el nuevo modelo de regresión lineal múltiple con datos depurados disminuye la multicolinealidad y su calidad predictiva es también muy aceptable con valor de r square ajustado de 0.8919 por encima del 85% establecido como punto de referencia.

Como segunda estrategia para tratar con la multicolinealidad, se normalizaron datos para construir tres modelos de regresión, un modelo lineal múltiple, y otros dos modelos Lasso y Ridge con la finalidad de que se generaran coeficientes más estableces penalizando la condición de multicolinealidad. Estos modelos también ofrecen calidad predictiva con valores de r square ajustado de 8931, 8931 y 8930 garantizando con ello que están por encima del topo establecido del 85% como métrica de calidad predictiva en este estadístico.

Todos los modelos de regresión lineal múltiple aquí construidos, cumplen con postulados de linealidad, homocedasticidad, normalidad e independencia de residuos destacando que el caso tuvo que tratar con el problema de multicolinealidad en las variables. En los modelos Lasso y Ridge por su estructura interna no aplican los postulados de linealidad y homocedasticidad pero si los de normalidad e independencia de residuos los cuales si cumplen con esta condición.

Al final se presenta las ecuaciones que generan las predicciones de cada modelo aquí construido, no se puede determinar y contestar a la pregunta ¿cuál modelo fue mejor?, de manera general, todos los modelos predictivos aquí referidos en este caso de estudio, ofrecen buena calidad predictiva y se identifica problemas de multicolinealidad en los datos y mecanismos para resolverlo.

El caso es una referencia académica para los lectores que les ayude por un lado identificar problemas de multicolinealidad en los datos y estrategias para poder resolver esta condición. Al final, el caso de estudio crea, evalúa y compara modelos de regresión multivariada.



